Support highlighting Markdown/RST in comments

taminomara opened the pull request:
This PR adds an optional feature that highlights Markdown, MySt (a Markdown extension used with Sphinx), and RST syntax in comments and provides Go To Definition implementation for cross-references. This feature is disabled by default; it can be enabled by specifying `doc.syntax` in `.emmyrc.json`.

<img width="1576" height="924" alt="image" src="https://github.com/user-attachments/assets/cb529517-06b4-4758-b1ee-d8f77dab2e18" />

# Caveats

- We use Semantic Tokens to provide syntax highlighting. Since there are no Semantic Tokens for bold and italic text, those are not highlighted.

CppCXY commented:
You just want to highlight the documentation in comments, so you only need to parse the comments once during semantic token request and highlight them in the appropriate places, rather than modifying the entire structure of the syntax tree.

taminomara commented:
Hm, I see your point. I can implement this. Although there will be complications if we want to support Go To Definition in comments.

I see it like this:

- make a cache for comment markup, in case if IDE requests semantic tokens twice;
- in Go To Definition request, if the user clicked inside of the documentation comment, we can check the comment markup cache to see if there's a cross reference in that place;
- in Document Selection Range request, we can either ignore comments, or use comment markup cache once again.

Would you agree with this implementation?

CppCXY commented:
Even if you re-parse everything, it will still be faster than using a cache.

taminomara commented:
Comment markup cache doesn't need its own AST, it will only save highlighted ranges.

taminomara commented:
Reworked this PR to avoid changes to AST.

CppCXY commented:
When the code block area uses Lua syntax, should we use luaLexer to parse it and highlight keywords, strings, and numbers?

taminomara commented:
> When the code block area uses Lua syntax, should we use luaLexer to parse it and highlight keywords, strings, and numbers?

Done:

<img width="634" height="288" alt="image" src="https://github.com/user-attachments/assets/b4d3a3b6-708d-4b79-ad20-9e63e96e0c3a" />

CppCXY commented:
<img width="604" height="282" alt="image" src="https://github.com/user-attachments/assets/ec615651-2d64-44d1-bdb2-424db242c7ca" />
Why does the highlighting look like this?

CppCXY commented:
If it works well, I plan to enable Markdown mode by default. This is how it currently performs in nvim. 
<img width="1216" height="789" alt="image" src="https://github.com/user-attachments/assets/3b2d7be2-b4d1-40eb-85e1-74c4b2bac869" />

What’s your opinion? @lewis6991

lewis6991 commented:
To begin with, I really don't think this should be enabled by default. I will use this feature, but based on conversations I've had with people, this will be polarising to a lot of people, especially if there are any issues, even minor ones. I suggest this is opt-in until we know it is stable.

- Does it only parse comments with `---` or all comments? If not I suggest this is limited to only `---` comments.

- Can we configure whether to use markdown or rst?

- This is a huge PR. Would it make sense to place the markdown and rst parser into separate crates?

taminomara commented on crates/emmylua_parser/src/lexer/lua_lexer.rs:
```diff
@@ -523,18 +552,26 @@ impl LuaLexer<'_> {
         }
     }
+
+    fn error<F, R>(&mut self, msg: F)
```
Not sure what you mean.

If it's about accepting a closure, it's just a micro optimization so that we don't run translation if error reporting is disabled.

taminomara commented:
> Yes, and I've also had someone say they want treesitter (which has more highlights and is faster) to handle this kind of highlighting, not the server.

Then there should be a way to disable highlighting even if `emmyrc` specifies otherwise. I think I'll do it as a separate PR though.

clason commented:
Please don't invent new off-spec options that only VS Code is set up to handle it. It's enough to make this opt-in via `.emmrc.json` (or the LSP configuration requests), anything beyond that makes it much more (and too) difficult to reason about.

clason commented:
I don't think the "active" LSP features need to be opt-in, as they need to be actively triggered on a documentation location anyway. (It would be different if references also included documentation locations -- which I feel they shouldn't either way.)

I still recommend the option split, since the syntax format is a project choice while the highlighting is a user choice. This needs to be respected. 

And I understand that the extension needs special support for this; that is fine. I amjust concerned with the server configuration not being driven by that -- everything should be (also) configurable through standard interfaces (config file and ConfigurationChanged notifications).

taminomara commented:
> I don't think the "active" LSP features need to be opt-in

There's also selection ranges, which is not an active feature. If we're worrying about stability, this should also be opt-in, at least for now.

Also, hover is not an "active" feature, but it's easy to trigger it accidentally. Should it be opt-in as well?

taminomara commented:
Added `semanticTokens.renderDocumentationMarkup`, default is `false`.

The reason for this name is for consistency with possible future options.

For example, currently we don't provide `documentation` modifier for semantic tokens related to documentation comments. In future, we might implement this, but we'll need to make it opt-out because at least in VSCode, most themes render documentation comments in horrendously bright colors.

> And I understand that the extension needs special support for this; that is fine. I amjust concerned with the server configuration not being driven by that -- everything should be (also) configurable through standard interfaces (config file and ConfigurationChanged notifications).

I've created #680 to address this issue. When it's merged, VSCode and NeoVim (and other editors) will be able to configure things using the same mechanism.

CppCXY commented on crates/emmylua_parser_desc/src/testlib.rs:
```diff
@@ -0,0 +1,70 @@
+    };
+    let ranges = parser.parse(code, LuaDocDescription::cast(desc).unwrap());
+    let result = format_result(code, ranges);
+    assert_eq!(result.trim(), expected.trim());
```
If assert is written inside the function, will the error location always be inside this function when an error occurs?

CppCXY commented on crates/emmylua_ls/src/handlers/test_lib/mod.rs:
```diff
@@ -195,29 +191,24 @@ impl ProviderVirtualWorkspace {
         );
         let Some(result) = result else {
-            return false;
+            panic!("completion result is empty");
```
If panic occurs inside the function, does that mean I can’t quickly jump to the exact error location when an error happens?

CppCXY commented on crates/emmylua_parser_desc/src/util.rs:
```diff
@@ -0,0 +1,504 @@
+    } else {
+        static RE: LazyLock<regex::Regex> =
+            LazyLock::new(|| regex::Regex::new(r"^\p{Ps}|\p{Pi}|\p{Pf}$").unwrap());
+        let mut tmp = [0; 4];
```
I don’t want to introduce regular expressions into the base library. Maybe you can use unicode-general-category instead.

CppCXY commented:
please donot change my `local` keyword color
now: 
<img width="328" height="111" alt="image" src="https://github.com/user-attachments/assets/f520c4f5-caba-4cf1-b1ad-2554f579afa2" />

before:
<img width="281" height="123" alt="image" src="https://github.com/user-attachments/assets/02dd1285-0a70-4209-98f7-3db8cab1a09d" />

CppCXY commented:
please resolve conflicts

CppCXY commented:
If everything works well, I will merge it before the release tomorrow, but please resolve the conflicts first.

taminomara commented:
Rebased.

Since you want to release tomorrow, I've also added support for bold/italic highlighting via emmy annotator, see also EmmyLua/VSCode-EmmyLua#248.
